source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -0,0 +1,595 @@
+  assert(fresh_ids.size() >= 2 && "Not enough fresh ids.");
+
+  // Update the module id bound
+  for (auto id : fresh_ids) {
+    fuzzerutil::UpdateModuleIdBound(ir_context, id);
+  }
+
+  // Create the block where the instruction is executed by splitting the
+  // original block.
+  auto execute_block = block->SplitBasicBlock(
+      ir_context, fresh_ids[0],
+      fuzzerutil::GetIteratorForInstruction(block, instruction));
+
+  // Create the merge block for the conditional that we are about to create by
+  // splitting execute_block (this will leave |instruction| as the only
+  // instruction in |execute_block|).
+  auto merge_block = execute_block->SplitBasicBlock(
+      ir_context, fresh_ids[1],
+      fuzzerutil::GetIteratorForInstruction(execute_block, next_instruction));
+
+  // Propagate the fact that the block is dead to the newly-created blocks.
+  if (transformation_context->GetFactManager()->BlockIsDead(block->id())) {
+    transformation_context->GetFactManager()->AddFactBlockIsDead(
+        execute_block->id());
+    transformation_context->GetFactManager()->AddFactBlockIsDead(
+        merge_block->id());
+  }
+
+  // Initially, consider the merge block as the alternative block to branch to
+  // if the instruction should not be executed.
+  auto alternative_block = merge_block;
+
+  // Add an unconditional branch from |execute_block| to |merge_block|.
+  execute_block->AddInstruction(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpBranch, 0, 0,
+      opt::Instruction::OperandList{
+          {SPV_OPERAND_TYPE_ID, {merge_block->id()}}}));
+
+  // If the instruction has a result id, we need to:
+  // - add an additional block where a dummy result is obtained by using the
+  // OpUndef instruction
+  // - change the result id of the instruction to a fresh id
+  // - add an OpPhi instruction, which will have the original result id of the
+  //   instruction, in the merge block.
+  if (instruction->HasResultId()) {
+    // We need 3 more fresh ids for 1 additional block and 2 additional
+    // instructions.
+    assert(fresh_ids.size() >= 5 && "Not enough fresh ids.");
+
+    // Create a new block using a fresh id for its label.
+    auto alternative_block_temp = MakeUnique<opt::BasicBlock>(
+        MakeUnique<opt::Instruction>(ir_context, SpvOpLabel, 0, fresh_ids[2],
+                                     opt::Instruction::OperandList{}));
+
+    // Keep the original result id of the instruction in a variable.
+    uint32_t original_result_id = instruction->result_id();
+
+    // Set the result id of the instruction to a fresh id.
+    instruction->SetResultId(fresh_ids[3]);
+
+    // Add an OpUndef instruction, with the same type as the original
+    // instruction and a fresh id, to the new block.
+    alternative_block_temp->AddInstruction(MakeUnique<opt::Instruction>(
+        ir_context, SpvOpUndef, instruction->type_id(), fresh_ids[4],
+        opt::Instruction::OperandList{}));
```

Comment:
This is related to #3716. If `instruction` is an `OpFunctionCall` to the function that doesn't return anything, we might not be able to create an `OpUndef` of `OpTypeVoid`. This means that we won't be able to create an `OpPhi` as well.

I think this should be OK if we prohibit all usages of void ids as well as `OpUndef` of void type. Otherwise (if we only prohibit `OpUndef`), we might have some trouble with `TransformationAddSynonym` that can create `OpCopyObject` of void type.

@paulthomson for info.

Reply:
Additionally, you could create an `OpUndef` in the global scope. This should reduce some amount of code.

Reply:
That could work. I do wonder if keeping things local is more interesting in this case though. Hard to know for sure. I worry that spirv-fuzz might generally add too many uses of globals.

I guess if we see a function call with a void return type and if the void result is used (which seems unlikely) we could give up trying to do this transformation.

Reply:
I have changed the code so that, whenever an instruction has side effects and returns a void type:
- is treated like an instruction that does not return anything if its result is not used in the module
- causes the transformation to be inapplicable otherwise

For now, the transformation still adds ``OpUndef`` locally. @paulthomson What advantages/disadvantages are there to have it global? Is there a transformation to add a global OpUndef?

Reply:
Sounds good! I suggest keeping it local for now. Replacing it with a global undef is an easy transformation that could occur later.